### PR TITLE
idris2 0.2.0 (new formula)

### PR DESCRIPTION
**Pull request by vladimyr**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

### CODE REVIEW

**Comment by SMillerDev** [comment]
Please indicate how this would fill our criteria for versioned formulas (https://docs.brew.sh/Versions) and why we might want to ship an older version?

**Comment by alebcay** [comment]
It appears that [`idris`](https://github.com/idris-lang/Idris-dev) is the older version and [`idris2`](https://github.com/idris-lang/Idris2) is a newer, separate project (from the same organization, but presumably with some design changes).

> Idris 2 is an early preview of the next generation of Idris, implemented in Idris.

> Idris 2 is mostly backwards compatible with Idris 1, with some minor exceptions. The most notable user visible differences, which might cause Idris 1 programs to fail to type check, are: [...]

**Comment by vladimyr** [comment]
> It appears that [`idris`](https://github.com/idris-lang/Idris-dev) is the older version and [`idris2`](https://github.com/idris-lang/Idris2) is a newer, separate project (from the same organization, but presumably with some design changes).
> 
> >

Exactly, Idris2 is mostly [backward compatible](https://idris2.readthedocs.io/en/latest/updates/updates.html) but there are some edge cases. For instance, the FFI example from the Idris HB test suite isn't compatible with Idris2. To me, it looks a lot like python2 vs python3 migration, less drastically tho.

**Comment by SMillerDev** [comment]
If its a completely different project it shouldn't be marked as versioned/keg-only.

File: Formula/idris@2.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+class IdrisAT2 < Formula
```
Comment (SMillerDev) [inline Formula/idris@2.rb:1]: The program name is `idris2`, which is different than "version 2 of idris"

File: Formula/idris@2.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  desc "Pure functional programming language with dependent types"
+  homepage "https://www.idris-lang.org/"
+  url "https://github.com/idris-lang/Idris2/archive/v0.2.0.tar.gz"
+  sha256 "cc7c48d3fbaa55e536890c9eb4701755081d2b5df9a65f0ecc3e00ee5ccf0a0f"
+
+  keg_only :versioned_formula
```
Comment (SMillerDev) [inline Formula/idris@2.rb:7]: it's a different formula, not versioned

File: Formula/idris@2.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+  depends_on "coreutils" => :build
+  depends_on "chezscheme"
+
+  def install
+    ENV.deparallelize
```
Comment (SMillerDev) [inline Formula/idris@2.rb:13]: Can't build in parallel?

File: Formula/idris@2.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    scheme = Formula["chezscheme"].bin/"chez"
+    system "make", "bootstrap", "SCHEME=#{scheme}", "PREFIX=#{libexec}"
+    system "make", "install", "PREFIX=#{libexec}"
+    idris2_app = libexec/"bin/idris2_app"
+    (bin/"idris2").write_env_script idris2_app/"idris2.so",
```
Comment (SMillerDev) [inline Formula/idris@2.rb:18]: shouldn't this point to the dylib?

File: Formula/idris@2.rb (hunk 3 above)
Comment (vladimyr) [reply Formula/idris@2.rb:13]: Can't, fails without it.

File: Formula/idris@2.rb (hunk 4 above)
Comment (vladimyr) [reply Formula/idris@2.rb:18]: It is counterintuitive but this is actually a binary. Idris2 outputs binaries with `.so` extension and that is not configurable.

Comment (SMillerDev) [reply Formula/idris@2.rb:18]: Could you make that a comment in the formula? Because someone in the future will think the same thing I just did

**Comment by vladimyr** [comment]
> If its a completely different project it shouldn't be marked as versioned/keg-only.

It depends on your definition of _different_. Yes Idris and Idris2 are developed inside different git repos but they are still part of the same org, developed by the same group of people and they share the official homepage that doesn't market them as different projects.

**Comment by SMillerDev** [comment]
> It depends on your definition of different.

That's true, in my experience homebrew only refers to versioned formulae if they share a codebase. I'll make a pull request to clarify that in documentation.

File: Formula/idris@2.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+                                    :LD_LIBRARY_PATH => "#{idris2_app}:#{ENV["LD_LIBRARY_PATH"]}"
+    lib.install_symlink Dir["#{libexec}/lib/*.dylib"]
+  end
+
+  test do
+    ENV.prepend_path "DYLD_LIBRARY_PATH", lib
```
Comment (vladimyr) [inline Formula/idris@2.rb:24]: @SMillerDev Is there a way to avoid this sorcery for unlinked formulae?

Comment (SMillerDev) [reply Formula/idris@2.rb:24]: Why would the formula be unlinked? It's installed before the test

Comment (vladimyr) [reply Formula/idris@2.rb:25]: It is keg only (in its current state) so libraries won't get linked to `/usr/local/lib`.

**Comment by vladimyr** [comment]
I've updated PR according to #7650

**Review by SMillerDev — DISMISSED** [review]
Thanks @vladimyr ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉

File: Formula/idris2.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
+class Idris2 < Formula
+  desc "Pure functional programming language with dependent types"
+  homepage "https://www.idris-lang.org/"
+  url "https://github.com/idris-lang/Idris2/archive/v0.2.0.tar.gz"
+  sha256 "cc7c48d3fbaa55e536890c9eb4701755081d2b5df9a65f0ecc3e00ee5ccf0a0f"
+
+  depends_on "coreutils" => :build
+  depends_on "chezscheme"
```
Comment (madushan1000) [inline Formula/idris2.rb:8]: Hi, @vladimyr shouldn't "chezscheme" be only a build time dependency, or is it needed for runtime code generation too? 
Thank you for your efforts, I'm eager to see this PR get merged :)

Comment (vladimyr) [reply Formula/idris2.rb:8]: It is used as Idris 2 compiler backend which makes it required in _runtime_. You are welcome, I'm doing it in appreciation for all the great things Idris folks created and thought us 😉

**Comment by madushan1000** [comment]
Ohh, I thought since idris2 is self hosting there might not be any other runtime dependencies, my bad. Thank you for taking the time to respond.

**Comment by samford** [comment]
The only failure here was the `GitHub repository too new (< 30 days old)` audit, since this is around 28 days old at this point.

```
==> brew audit --skip-style idris2 --new-formula
==> FAILED
Error: 1 problem in 1 formula detected
idris2:
  * GitHub repository too new (<30 days old)
```

I went ahead and merged this, per @SMillerDev's request.

**Comment by vladimyr** [comment]
> The only failure here was the `GitHub repository too new (< 30 days old)` audit, since this is around 28 days old at this point.

Yeah, it is a re-implementation of well-known language with new compiler backend and in the new repo so it didn't fit into usual constraints 🤷
